thirdparty/SA/libdivsufsort/lib/utils.c

```diff
@@ -0,0 +1,382 @@
+/*
+ * utils.c for libdivsufsort
+ * Copyright (c) 2003-2008 Yuta Mori All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person
+ * obtaining a copy of this software and associated documentation
+ * files (the "Software"), to deal in the Software without
+ * restriction, including without limitation the rights to use,
+ * copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following
+ * conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+//#include "divsufsort_private.h"
+#include "/home/z/zd4/SA/libdivsufsort/include/divsufsort_private.h"
```

Comment:
This line seems to be one of the things that's breaking the CI tests due to the absolute path not existing on the test systems.  My guess is that the commented out line is how you originally wanted this to work, but that you didn't know how to feed a `-I` flag to point at the right place, so replaced it with an absolute path?  If that's so, then I _believe_ the fix would simply be to add a `-I` pointing to the third-party directory to the `chpl` compilation line in the Makefile (this might also permit you to remove some of the relative paths in your `require` statements?  I'll do a quick check to ascertain that this is the case.

Reply:
Yep, passing a `-I` flag to the compiler should resolve the need to put an absolute path in here.  Adding it to CHPL_FLAGS towards the top of the main `Makefile` should permit you to revert this to the commented out `#include` line.

Correcting one thing I said above: Doing so should permit you to remove the relative paths in your `require` statements for header files, but not for `.c` files.

Here's my toy program demonstrating this:

tmp-thirdparty/foo.c:
```chapel
#include "foo.h"

int foo() {
  return 42;
}
```

tmp-thirdparty/foo.h:
```chapel
int foo(void);
```

testfoo.chpl:
```chapel
require "foo.h";
require "tmp-thirdparty/foo.c";

use SysCTypes;

extern proc foo(): c_int;

writeln(foo());
```

```bash
$ chpl testit.chpl -I tmp-thirdparty
```

Reply:
Thanks for your suggestion!
I am a little bit curious on how this happens. At first I use the absolute path and then I replace them with a relative path. You may check my src/SACA.chpl file. I do not know why the CI tests can still find my original verion.

Reply:
The error I was looking at was [this one](https://github.com/mhmerrill/arkouda/runs/1554818429?check_suite_focus=true#step:5:205) (do you get to see the log as an external developer?  Hope so), where the error is happening during the C compilation step for `divsufsort.c`:

```> /__w/arkouda/arkouda//src/../thirdparty/SA/libdivsufsort/lib/divsufsort.c:28:10: fatal error: /home/z/zd4/SA/libdivsufsort/include/divsufsort_private.h: No such file or directory
>  #include "/home/z/zd4/SA/libdivsufsort/include/divsufsort_private.h"
>           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> compilation terminated.
> make: *** [/tmp/chpl-root-3021.deleteme/Makefile:55: /tmp/chpl-root-3021.deleteme/divsufsort.c.o] Error 1
> error: compiling generated source
```

Essentially what happens is:  Any `.c` files specified by `require` statements or on the `chpl` command line are compiled and linked into the code generated by the `chpl` compiler.  So during this C compilation step, the CI process is going to look for the `.h` file at its absolute path and not finding it.  Any `-I` flags passed to the `chpl` command line are also passed to these `.c` compilation steps, and so can be used to specify the path to the header.

You can potentially get a better sense of what is happening if you add the `--print-commands` flag to the `chpl` compilation line, which will show the individual compile and link steps that take place to make the binary.

Reply:
> Yep, passing a `-I` flag to the compiler should resolve the need to put an absolute path in here. Adding it to CHPL_FLAGS towards the top of the main `Makefile` should permit you to revert this to the commented out `#include` line.
> 
> Correcting one thing I said above: Doing so should permit you to remove the relative paths in your `require` statements for header files, but not for `.c` files.
> 
> Here's my toy program demonstrating this:
> 
> tmp-thirdparty/foo.c:
> 
> ```chapel
> #include "foo.h"
> 
> int foo() {
>   return 42;
> }
> ```
> 
> tmp-thirdparty/foo.h:
> 
> ```chapel
> int foo(void);
> ```
> 
> testfoo.chpl:
> 
> ```chapel
> require "foo.h";
> require "tmp-thirdparty/foo.c";
> 
> use SysCTypes;
> 
> extern proc foo(): c_int;
> 
> writeln(foo());
> ```
> 
> ```shell
> $ chpl testit.chpl -I tmp-thirdparty
> ```

---
I just realize that I only change the path in src/SACA.chpl and forgot the other files.
Thank you so much!